packages/core-utils/src/browser-storage.ts

```diff
@@ -0,0 +1,174 @@
+ * Values must be JSON-serializable. The 'dynamic' version is preferred if possible, but dynamic
+ * values might be deleted in case of space overflow. For critical data, use the 'static' version.
+ */
+
+class InMemoryStorage {
+  storage = Object.create(null);
+
+  get length() {
+    return _.size(this.storage);
+  }
+
+  key(index: any) {
+    return Object.keys(this.storage)[index] || null;
+  }
+
+  getItem(key: any) {
+    return _.has(key, this.storage) ? _.get(key, this.storage) : null;
+  }
+
+  setItem(key: any, value: any) {
+    this.storage[key] = value;
+  }
+
+  removeItem(key: any) {
+    delete this.storage[key];
+  }
+
+  clear() {
+    this.storage = Object.create(null);
+  }
+}
+
+export const getLocalStorage = _.once(() => {
+  try {
+    return window.localStorage;
+  } catch (error) {
+    return new InMemoryStorage();
+  }
+});
+
+export const getSessionStorage = _.once(() => {
+  try {
+    return window.sessionStorage;
+  } catch (error) {
+    return new InMemoryStorage();
+  }
+});
+
+const forceSetItem = (storage: any, key: any, value: any) => {
+  while (true) {
+    try {
+      storage.setItem(key, value);
+      return;
+    } catch (error) {
+      const candidates = _.filter(([k]: [any]) => _.startsWith('dynamic-storage/', k), _.toPairs(storage));
+      if (!candidates.length) {
+        console.error('Could not write to storage, and no entries to delete');
+        return;
+      }
+      const [chosenKey] = _.head(
+        _.sortBy(([_k, v]: [any, any]) => {
+          const data: any = maybeParseJSON(v);
+          return data && _.isInteger(data.timestamp) ? data.timestamp : -Infinity;
+        }, candidates)
+      ) as [any];
+      storage.removeItem(chosenKey);
+    }
+  }
+};
+
+export const getStatic = (storage: any, key: any) => {
+  return maybeParseJSON(storage.getItem(key));
+};
+
+/**
+ * Sets a key/value in browser storage, or removes the key from storage if `value` is undefined.
+ *
+ * Static storage will not be automatically deleted in the case of space overflow. If your storage usage
+ * is transient (for example, navigating between routes and preserving data where the loss of it is not
+ * critical), instead use `setDynamic`.
+ *
+ * See also `staticStorageSlot`, which can be passed to the `useStore` hook function.
+ *
+ * @param storage an object implementing the Storage interface
+ * @param key the key for storing the value
+ * @param value the new JSON-serializable value to be stored, or `undefined` to remove the key.
+ */
+export const setStatic = (storage: any, key: any, value: any) => {
+  if (value === undefined) {
+    storage.removeItem(key);
+  } else {
+    forceSetItem(storage, key, JSON.stringify(value));
+  }
+};
+
+export const listenStatic = (storage: any, key: any, fn: any) => {
+  window.addEventListener('storage', (e) => {
+    if (e.storageArea === storage && e.key === key && e.newValue) {
```

Comment:
Added check for e.newValue, because maybeParseJSON expects a string (not null or undefined).

However, since this was being called in Javascript before (without type checks), I am using undefined in the else block, because that's what maybeParseJSON's catch block does.

Reply:
I think this check could be removed if we have the correct type on e, window event 'storage' must have a documented event type that will give safety for newValue attribute and more.